core/keyserver/paginator.go

```diff
@@ -60,77 +58,46 @@ func (s SourceMap) ParseToken(token string) (*rtpb.ReadToken, error) {
 }
 
 // First returns the first read parameters for this source.
-func (s SourceMap) First() *rtpb.ReadToken {
+func (s SourceList) First() *rtpb.ReadToken {
 	if len(s) == 0 {
 		// Empty struct means there is nothing else to page through.
 		return &rtpb.ReadToken{}
 	}
-	firstLog := int64(math.MaxInt64)
-	for logID := range s {
-		if logID < firstLog {
-			firstLog = logID
-		}
-	}
 	return &rtpb.ReadToken{
-		ShardId:      firstLog,
-		LowWatermark: s[firstLog].LowestWatermark,
+		SliceIndex:   0,
+		LowWatermark: s[0].LowestWatermark,
 	}
 }
 
 // Next returns the next read token. Returns an empty struct when the read is finished.
 // lastRow is the (batchSize)th row from the last read, or nil if fewer than
 // batchSize + 1 rows were returned.
-func (s SourceMap) Next(rt *rtpb.ReadToken, lastRow *mutator.LogMessage) *rtpb.ReadToken {
+func (s SourceList) Next(rt *rtpb.ReadToken, lastRow *mutator.LogMessage) *rtpb.ReadToken {
 	if lastRow != nil {
 		// There are more items in this shard.
 		return &rtpb.ReadToken{
-			ShardId:      rt.ShardId,
+			SliceIndex:   rt.SliceIndex,
 			LowWatermark: lastRow.ID,
 		}
 	}
 
 	// Advance to the next shard.
-	nextShard, ok := s.NextShard(rt.ShardId)
+	nextShard, ok := s.NextShard(rt.SliceIndex)
 	if !ok {
 		return &rtpb.ReadToken{} // Encodes to ""
 	}
 	return &rtpb.ReadToken{
-		ShardId:      nextShard,
+		SliceIndex:   nextShard,
 		LowWatermark: s[nextShard].LowestWatermark,
 	}
 }
 
-// NextShard returns the next shardID from the SourceMap.
-// Returns false if there are no more shards or shardID is not in SourceMap.
-func (s SourceMap) NextShard(shardID int64) (int64, bool) {
-	// Sorted list of shardIDs.
-	shardIDs := sortedKeys(s)
-
-	// Index of current shard.
-	i := sort.Search(len(shardIDs), func(i int) bool { return shardIDs[i] >= shardID })
-	if i == -1 {
-		// shardID isn't in SourceMap.
-		return 0, false
-	}
-	if i >= len(shardIDs)-1 {
+// NextShard returns the next shardID from the SourceList.
+// Returns false if there are no more shards or shardID is not in SourceList.
+func (s SourceList) NextShard(shardID int64) (int64, bool) {
+	if shardID >= int64(len(s))-1 {
 		// there are no more shards to iterate over.
 		return 0, false
 	}
-	return shardIDs[i+1], true
+	return shardID + 1, true
```

Comment:
Would it be okay to replace the whole body with the following?
`return sliceIndex+1, sliceIndex+1 < len(s)`

Presumably the caller checks the `bool` before using the returned index?

Reply:
Yes, I thought about that, but other people on our team feel strongly about returning `0` or `nil` on error conditions, so I left the if statement in place.